### PR TITLE
feat: CRUD API for Taxonomy Tags [FC-0036]

### PR DESCRIPTION
**Pull request by yusuf-musleh**
### Description

This PR introduces the CRUD python and REST APIs for interacting with Taxonomy Tags. Creating, updating and deleting them.

### Supporting Information

Related Tickets:
- Closes https://github.com/openedx/modular-learning/issues/136

### Testing Instructions

Make sure all the tests pass, and they cover all the cases.

---
Private-ref: [FAL-3519](https://tasks.opencraft.com/browse/FAL-3519)

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @yusuf-musleh!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


This is currently a draft pull request. When it is ready for our review and all tests are green, click "Ready for Review", or remove "WIP" from the title, as appropriate.

File: openedx_tagging/core/tagging/models/base.py
Comment (pomegranited) [inline openedx_tagging/core/tagging/models/base.py:374]: We should ensure that the `parent` tag is in this taxonomy, e.g.

```suggestion
        if parent_tag_id:
            parent = self.tag_set.get(id=parent_tag_id)
```

File: openedx_tagging/core/tagging/models/base.py
Comment (pomegranited) [inline openedx_tagging/core/tagging/models/base.py:400]: Similarly here, we should catch the possible `Tag.DoesNotExist`, and ensure the view responds with a 404.

File: openedx_tagging/core/tagging/models/base.py
Comment (pomegranited) [reply openedx_tagging/core/tagging/models/base.py:374]: Also, we should also catch the `Tag.DoesNotExist` that this might throw and ensure the explanation gets back to the user [when calling the view](https://github.com/openedx/openedx-learning/pull/96/files#diff-215d20c663dcd5727c14740a72d09fe6fd22e3a92fd08db46036e2e799699917R549).

**Review by pomegranited — CHANGES_REQUESTED** [review]
This is looking good @yusuf-musleh !
Let me know when you're ready for a full review?

File: openedx_tagging/core/tagging/models/base.py
Comment (yusuf-musleh) [reply openedx_tagging/core/tagging/models/base.py]: Thanks for catching this! I updated it and added a test for this case, I initially was only testing for "invalid" tag ids forgot to check for the case of "valid but for other taxonomy".

File: openedx_tagging/core/tagging/models/base.py
Comment (yusuf-musleh) [reply openedx_tagging/core/tagging/models/base.py:400]: Addressed as well, thanks!

File: openedx_tagging/core/tagging/rest_api/v1/views.py
Comment (pomegranited) [inline openedx_tagging/core/tagging/rest_api/v1/views.py:656]: I just noticed that the [api exports `TagDoesNotExist = Tag.DoesNotExist`](https://github.com/openedx/openedx-learning/blob/d44775d83e00e642c96077da93d6ed82c06204be/openedx_tagging/core/tagging/api.py#L22), so maybe we should use that instead?

File: openedx_tagging/core/tagging/rest_api/v1/views.py
Comment (pomegranited) [inline openedx_tagging/core/tagging/rest_api/v1/views.py:482]: @yusuf-musleh I think your best approach would be to modify (and rename) [`TagListPermissions`](https://github.com/openedx/openedx-learning/blob/d44775d83e00e642c96077da93d6ed82c06204be/openedx_tagging/core/tagging/rest_api/v1/permissions.py#L38) to provide a `perm_map` like we did with [`ObjectTagObjectPermissions`](https://github.com/openedx/openedx-learning/blob/d44775d83e00e642c96077da93d6ed82c06204be/openedx_tagging/core/tagging/rest_api/v1/permissions.py#L23).

That will enforce the `oel_tagging.view_tag`, `oel_tagging.change_tag`, and `oel_tagging.delete_tag` rules before invoking the view methods.

But I think you're correct in your question -- these `oel_tagging.*_tag` rules need to be updated to reflect the changes to the taxonomy view/change rules made by https://github.com/openedx/openedx-learning/pull/94.

I think something like this would be sufficient?

```
def can_view_tag(user: UserType, tag: Tag | None = None) -> bool:
    """
    Users can view tags for any taxonomy they can view.
    """"
    taxonomy = tag.taxonomy.cast() if (tag and tag.taxonomy) else None
    return  user.has_perm(
        "oel_tagging.view_taxonomy",
        taxonomy,
    )

def can_change_tag(user: UserType, tag: Tag | None = None) -> bool:
    """
    Users can change tags for any taxonomy they can modify.
    """"
    taxonomy = tag.taxonomy.cast() if (tag and tag.taxonomy) else None
    return  user.has_perm(
        "oel_tagging.change_taxonomy",
        taxonomy,
    )

rules.add_perm("oel_tagging.add_tag", can_change_tag)
rules.add_perm("oel_tagging.change_tag", can_change_tag)
rules.add_perm("oel_tagging.delete_tag", can_change_tag)
rules.add_perm("oel_tagging.view_tag", can_view_tag)
rules.add_perm("oel_tagging.list_tag", can_view_tag)
```

File: tests/openedx_tagging/core/tagging/test_views.py
Comment (pomegranited) [inline tests/openedx_tagging/core/tagging/test_views.py:1208]: Once you've fixed the `TaxonomyTagsView.permission_classes`, you'll need to change your authentication for these tests to `self.staff`. But it's worth leaving in a few checks to ensure that `self.user` is not allowed to create or update tags.

**Review by pomegranited — CHANGES_REQUESTED** [review]
Your view changes are looking great @yusuf-musleh , and you're correct, there's some rules/permission changes needed.

File: openedx_tagging/core/tagging/rest_api/v1/views.py
Comment (yusuf-musleh) [reply openedx_tagging/core/tagging/rest_api/v1/views.py]: Sure thing, updated

File: openedx_tagging/core/tagging/rest_api/v1/views.py
Comment (yusuf-musleh) [reply openedx_tagging/core/tagging/rest_api/v1/views.py]: @pomegranited Thanks for the pointers! I went ahead with what you suggested, however I faced a few issues and took me quite some time to figure them out:
1. The `get_queryset` method in `TaxonomyTagsView` was initially returning a list, as it does some custom population of an attribute `subtags` and later serialized using the `TagsForSearchSerializer`. For the updated permissions (using `perm_map`) to work,`get_queryset` needs to return a queryset as it is used internally by DRF. The current solution I have (though not a huge fan of it) is getting the result list and constructing a queryset using `filter` and then repopulating the `subtags` again if applicable. Ideally, I think the approach should be to refactor the underlying implementation of [get_matching_tags](https://github.com/openedx/openedx-learning/blob/ca6bc859f69d5508d5457f42a61e8fd3263a7b53/openedx_tagging/core/tagging/rest_api/v1/views.py#L468) to use querysets instead. I couldn't get to that today, but I thought I'd mention it here to get your thoughts.
2. After adding the `perms_map` for `TagObjectPermissions`, I left the `has_object_permission` method in place, since it's a special case for the GET list endpoint. There is not `tag` object in that case, we have the Taxonomy object, so we utilize the `can_view_taxonomy` permission for the `oel_tagging.list_tag` case. It gets [called](https://github.com/openedx/openedx-learning/blob/ca6bc859f69d5508d5457f42a61e8fd3263a7b53/openedx_tagging/core/tagging/rest_api/v1/views.py#L429) in the `get_taxonomy` method to and handle whether the use can view the Tags under this Taxonomy or not.

File: openedx_tagging/core/tagging/api.py
Comment (bradenmacdonald) [inline openedx_tagging/core/tagging/api.py:132]: What's the advantage of making this a static method on `ObjectTag` rather than leaving it here in the API? I kind of prefer having it here since it forces usage of `api`, and we don't want others directly calling methods on `ObjectTag`. Though perhaps we can use the [isolated apps linter](https://github.com/openedx/edx-platform/blob/febcccc14767378492a7a897b66023dafafda619/setup.cfg#L163-L173) to enforce that in the future as well.

I don't have a strong opinion on this, so no need to change it if you think this makes more sense.

File: openedx_tagging/core/tagging/rest_api/v1/views.py
Comment (pomegranited) [reply openedx_tagging/core/tagging/rest_api/v1/views.py]: @yusuf-musleh 

> 1... For the updated permissions (using `perm_map`) to work, `get_queryset` needs to return a queryset as it is used internally by DRF... Ideally, I think the approach should be to refactor the underlying implementation of [get_matching_tags](https://github.com/openedx/openedx-learning/blob/ca6bc859f69d5508d5457f42a61e8fd3263a7b53/openedx_tagging/core/tagging/rest_api/v1/views.py#L468) to use querysets instead.

I agree.. have you seen @bradenmacdonald 's  https://github.com/openedx/openedx-learning/pull/92 ? I think that PR does what you need already. We ran out of budget to complete it, but if it's needed for this work, then maybe we can complete it here?

And what you did for the permissions looks great, thank you.

File: openedx_tagging/core/tagging/api.py
Comment (pomegranited) [reply openedx_tagging/core/tagging/api.py:328]: Ooh good points, @bradenmacdonald -- I agree that using `value` where we can makes sense.

File: openedx_tagging/core/tagging/api.py
Comment (pomegranited) [reply openedx_tagging/core/tagging/api.py:132]: I think @yusuf-musleh  did this so he could call `resync` from the `Taxonomy.delete_tags` and `Taxonomy.update_tags` model methods without creating a circular dependency.

But instead, we could leave `resync_object_tags` as-is, and call it from the api `update_tag_in_taxonomy` and `delete_tags_from_taxonomy` methods instead.

What do you think, @yusuf-musleh ?

Comment (pomegranited) [reply openedx_tagging/core/tagging/api.py:132]: Also, `resync_object_tags` could take a `taxonomy` as its first argument instead of the ObjectTags queryset, since we're currently only resyncing object tags for single taxonomy.

**Review by pomegranited — CHANGES_REQUESTED** [review]
We're getting there.. I've left a couple of minor comments, but would like to re-review once you've addressed Braden's suggestions.

File: openedx_tagging/core/tagging/rest_api/v1/permissions.py
Comment (yusuf-musleh) [reply openedx_tagging/core/tagging/rest_api/v1/permissions.py:57]: oh interesting! Thanks for catching this, I managed to reproduce and applied your suggested fix.

File: openedx_tagging/core/tagging/rules.py
Comment (yusuf-musleh) [reply openedx_tagging/core/tagging/rules.py]: oops I forgot to change it back when debugging, updated it.

File: openedx_tagging/core/tagging/api.py
Comment (yusuf-musleh) [reply openedx_tagging/core/tagging/api.py]: I initially moved the logic of `resync_object_tags` to be a static method in the model to avoid duplicate code and to avoid calling api functions from inside the model's methods (like `update_tags`). However I like @pomegranited's suggestion,  makes more sense, I took out the resync logic and placed in in the api and brought back the `resync_object_tags` to the api, removing the static method: https://github.com/openedx/openedx-learning/commit/bdaffeb0d0e718ce56321be6db9a64b83e64bc47

PS: I actually forgot to do the resync of object tags in the delete method, I just added it to the `delete_tags_from_taxonomy`. Thanks for the reminder!

File: openedx_tagging/core/tagging/api.py
Comment (yusuf-musleh) [reply openedx_tagging/core/tagging/api.py]: I see, that makes sense, I updated the implementation to use values instead of IDs. https://github.com/openedx/openedx-learning/commit/3abce26a6fdd11f89dd4493db79a4d41a008c7c3

File: openedx_tagging/core/tagging/rest_api/v1/views.py
Comment (yusuf-musleh) [reply openedx_tagging/core/tagging/rest_api/v1/views.py]: @pomegranited I went through @bradenmacdonald 's https://github.com/openedx/openedx-learning/pull/92, lots of great work there, and already does alot of the heavy lifting for the refactoring, however since there are substantial changes made in that PR (and some more work remaining there) I feel it might get pretty messy to include it as part of this PR. Would it make sense to leave the current code in this PR as is and then continue working on https://github.com/openedx/openedx-learning/pull/92 and along with the refactoring needed here as part of a separate ticket? 

Alternatively, I could copy the relevant changes from that https://github.com/openedx/openedx-learning/pull/92 and include it as part of this PR? What do you guys think?

Comment (bradenmacdonald) [reply openedx_tagging/core/tagging/rest_api/v1/views.py]: > Would it make sense to leave the current code in this PR as is and then continue working on https://github.com/openedx/openedx-learning/pull/92 and along with the refactoring needed here as part of a separate ticket?

Yes, I think that makes sense :)

File: tests/openedx_tagging/core/tagging/test_views.py
Comment (pomegranited) [inline tests/openedx_tagging/core/tagging/test_views.py:1212]: Your view tests are great @yusuf-musleh , thank you! Very readable.

**Review by pomegranited — CHANGES_REQUESTED** [review]

File: openedx_tagging/core/tagging/rest_api/v1/views.py
Comment (pomegranited) [reply openedx_tagging/core/tagging/rest_api/v1/views.py]: Created https://github.com/openedx/modular-learning/issues/142 for this.

File: openedx_tagging/core/tagging/rest_api/v1/serializers.py
Comment (yusuf-musleh) [reply openedx_tagging/core/tagging/rest_api/v1/serializers.py:260]: The implementation in the `Taxonomy.delete_tags` is actually an incorrect one, since the simple `value__in` filter doesnt account for case-insensitve values, the current tests didn't catch it. Thanks for pointing it out, I'll update it.

> Also, do we need to this extra validation in these serializers, since the implementation also enforces the same rules?

My initial thought process was to add the validation in the api incase the functions were called directly rather than only in the rest_api/views (through serializers), however then I figured we could also make use of the validation that DRF provides in addition to that.

The rules they enforce are _slightly_ different, you'll notice the validation that happens in the serializer checks if the Tag values provided are valid (i.e exist in the DB regardless to which taxonomy they belong to) however the validation that happens in the api method checks if the **Taxonomy** has the provided (Tag) values, slight difference where the former would return a 400 and the later would return a 404. I guess the end result is technically the same (the request fails), just thought this give more granularity in checks and errors.

File: openedx_tagging/core/tagging/rest_api/v1/views.py
Comment (yusuf-musleh) [reply openedx_tagging/core/tagging/rest_api/v1/views.py]: Great, thanks for creating it @pomegranited !

File: openedx_tagging/core/tagging/rest_api/v1/serializers.py
Comment (yusuf-musleh) [reply openedx_tagging/core/tagging/rest_api/v1/serializers.py:260]: > The implementation in the Taxonomy.delete_tags is actually an incorrect one, since the simple value__in filter doesnt account for case-insensitve values, the current tests didn't catch it.

Actually, doing some tests, surprising the simple implementation works 😮, It might be because of the special custom `case_insensitive_char_field` we use in the model? Since that is generally not the default behavior for django's charfields. In any case, that was a pleasant surprise!

File: openedx_tagging/core/tagging/api.py
Comment (yusuf-musleh) [reply openedx_tagging/core/tagging/api.py]: Got it, I knew there was a reason I didnt add the resync initially to the delete tags functionality 😅 . And thanks for pointing out the resync that needs to happen when creating new tag. I've updated it with the changes here: https://github.com/openedx/openedx-learning/pull/96/commits/dd35747d6f4d7f4c11057800a206a69186948890

File: openedx_tagging/core/tagging/rest_api/v1/serializers.py
Comment (pomegranited) [reply openedx_tagging/core/tagging/rest_api/v1/serializers.py]: Ah lovely, I'm glad to hear that case-insensitive fields aren't so hard to interact with!

I'm still not convinced that the extra Serializer validation is needed, since the model does what it does and more. But I'll leave that to you & upstream's discretion.

**Review by pomegranited — APPROVED** [review]
:+1: Looks great to me @yusuf-musleh ! Ready for your eyes @bradenmacdonald .

Note: needs a version bump before merging.

- [x] I tested this using the django devserver and the DRF REST API UI.
- [x] I read through the code -- great tests!
- [x] <strike>I checked for accessibility issues</strike>
- [x] Includes documentation
- [x] Commit structure follows [OEP-0051](https://github.com/edx/open-edx-proposals/pull/189)

File: openedx_tagging/core/tagging/rest_api/v1/serializers.py
Comment (bradenmacdonald) [inline openedx_tagging/core/tagging/rest_api/v1/serializers.py:199]:
```suggestion
    Serializer of the body for the Taxonomy Tags CREATE request
```
If you say "view", it's unclear if this is for the request body or the response body.

File: openedx_tagging/core/tagging/rest_api/v1/serializers.py
Comment (bradenmacdonald) [inline openedx_tagging/core/tagging/rest_api/v1/serializers.py:204]: Is this `source` actually doing anything?

File: openedx_tagging/core/tagging/rest_api/v1/serializers.py
Comment (bradenmacdonald) [inline openedx_tagging/core/tagging/rest_api/v1/serializers.py:214]: Personally I would put this validation in the `Taxonomy.add_tag` method, where most other validation is anyways. That way it gets used for both the python API and REST API. Currently, only the REST API has this validation. You can leave it as-is though.

Removing the validation method(s) from these three serializers would also make the serializer code much more compact.

File: openedx_tagging/core/tagging/rest_api/v1/views.py
Comment (bradenmacdonald) [inline openedx_tagging/core/tagging/rest_api/v1/views.py:435]:
```suggestion
        * id (required) - The ID of the taxonomy to create a Tag for
```

(very minor nit:) Using `pk` to refer to the primary key is kind of an internal django thing that REST API integration developers may not be familiar with. I would simply say `ID` in the documentation strings.

File: openedx_tagging/core/tagging/rest_api/v1/views.py
Comment (bradenmacdonald) [inline openedx_tagging/core/tagging/rest_api/v1/views.py:445]:
```suggestion
        POST api/tagging/v1/taxonomy/:id/tags                                       - Create a Tag in taxonomy
```

File: openedx_tagging/core/tagging/rest_api/v1/views.py
Comment (bradenmacdonald) [inline openedx_tagging/core/tagging/rest_api/v1/views.py:475]: Nit: I think it would be a bit more RESTful to put the old value in the URL and just use `value` as the name of the new value field. Though since `value` is a mutable ID it's not a huge deal. So no need to change for now.

What I _would_ definitely recommend we do now is remove the `PUT` example from the docstring. Because if we add the ability to change other fields like `external_id` via this API endpoint in the future, anyone who is using `PUT` to change the name will now be accidentally overwriting/deleting the `external_id`, since PUT means "completely replace the tag with this new spec". But `PATCH` explicitly means "update the fields I specify to the new values and leave others unchanged".

**Review by bradenmacdonald — COMMENTED** [review]
Almost there! You don't have to do all of these as changes, just let me know your thoughts.

See also my suggestion on the edx-platform PR about changing how we use `reverse` so that this API can be hosted in any namespace, not just `oel_tagging`

File: openedx_tagging/core/tagging/rest_api/v1/serializers.py
Comment (yusuf-musleh) [reply openedx_tagging/core/tagging/rest_api/v1/serializers.py]: Got it, updated.

File: openedx_tagging/core/tagging/rest_api/v1/views.py
Comment (yusuf-musleh) [reply openedx_tagging/core/tagging/rest_api/v1/views.py]: Makes sense, updated.

File: openedx_tagging/core/tagging/rest_api/v1/views.py
Comment (yusuf-musleh) [reply openedx_tagging/core/tagging/rest_api/v1/views.py:470]: I see, for sure, I removed the `PUT` example as suggested, and kept the endpoint as is for now.

File: openedx_tagging/core/tagging/rest_api/v1/serializers.py
Comment (yusuf-musleh) [reply openedx_tagging/core/tagging/rest_api/v1/serializers.py]: After trying it out, it's definitely more concise and having the validations all in one place is cleaner, thanks @pomegranited and @bradenmacdonald for the suggestions! I updated it.

File: openedx_tagging/core/tagging/rest_api/v1/serializers.py
Comment (yusuf-musleh) [reply openedx_tagging/core/tagging/rest_api/v1/serializers.py]: I removed it and it doesn't seem like it had any impact. I think this is used more for serializing model/response data rather than handing request data.

**Comment by yusuf-musleh** [comment]
@bradenmacdonald Thanks for the review! I address/applied your suggestions, including the name space suggestion, I also bumped the version, it should hopefully be good to go now.

File: openedx_tagging/core/tagging/models/base.py
Comment (bradenmacdonald) [inline openedx_tagging/core/tagging/models/base.py:375]: I don't understand why we even need these lines ^^. Isn't the check below, which raises `Tag.DoesNotExist` sufficient? This just does an extra database query for no reason. We don't really care if the value exists already but in a different taxonomy - the check below already accounts for "value exists in another taxonomy" and "value doesn't exist at all"

**Review by bradenmacdonald — APPROVED** [review]
Thanks for those changes! This is good to merge, but I would like to see those duplicate checks removed first, unless there's some reason to keep them in that I'm missing.

File: openedx_tagging/core/tagging/models/base.py
Comment (yusuf-musleh) [reply openedx_tagging/core/tagging/models/base.py]: It is indeed sufficient, I just thought it adds a bit more granularity and description to the errors for the end user. However, you're right, it seems that the extra DB query for this might not be worthwhile. I've removed all of them and updated some test cases.

File: openedx_tagging/core/tagging/models/base.py
Comment (yusuf-musleh) [reply openedx_tagging/core/tagging/models/base.py]: Yup, addressed above.

File: openedx_tagging/core/tagging/models/base.py
Comment (yusuf-musleh) [reply openedx_tagging/core/tagging/models/base.py]: Similarly, removed and addressed above.

**Comment by yusuf-musleh** [comment]
@bradenmacdonald Thanks for the review! I've updated the PR to remove the duplicate checks, and the version bump is already included.

**Comment by openedx-webhooks** [comment]
@yusuf-musleh 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/openedx-learning&entry.867055334=https://github.com/openedx/openedx-learning/pull/96&entry.1484655318=https://github.com/yusuf-musleh&entry.752974735=2023-10-11+14:51&entry.1917517419=2023-10-22+23:07&entry.2133058324=Yes) so we can improve your experience in the future.

**Pull request merged**
